From 98ed5b3eade6ac8b16f469dce822f4b19b903d9f Mon Sep 17 00:00:00 2001
From: Tal Yacobi <tyacobi@habana.ai>
Date: Tue, 13 May 2025 11:17:22 +0300
Subject: [PATCH] [SW-221571] drivers: defer init of device drivers

Some devices are not supported in pcie card type
(e.g., in HL-338).

Defer init of these devices to runtime
instead of checking if they're supported in the driver.

Change-Id: I517c01311e4908ad6a91b60cdf48392db1982d11
Signed-off-by: Tal Yacobi <tyacobi@habana.ai>
---

diff --git a/boards/hl/gaudi2/hl_gaudi2_mgmt_asic.dts b/boards/hl/gaudi2/hl_gaudi2_mgmt_asic.dts
index 06b2b5d..6e60f51 100644
--- a/boards/hl/gaudi2/hl_gaudi2_mgmt_asic.dts
+++ b/boards/hl/gaudi2/hl_gaudi2_mgmt_asic.dts
@@ -111,6 +111,7 @@
 		compatible = "ti,tmp464";
 		reg = <0x48>;
 		node_label = "TMP464";
+		zephyr,deferred-init;
 	};
 
 	machxox: machxox@25 {
@@ -128,6 +129,7 @@
 		reg = <0x25>;
 		led-ctrl-reg = <0x40>;
 		node_label = "HL288_LED_CPLD";
+		zephyr,deferred-init;
 
 		led0: led_0 {
 			node_label = "QSFP1_LED_Y";
@@ -174,6 +176,7 @@
 		reg = <0x41>;
 		node_label = "XDPP_1100";
 		bus = <&i2c1>;
+		zephyr,deferred-init;
 	};
 
 	xdpe132g5h: xdpe132g5h@30 {
@@ -182,6 +185,7 @@
 		reg-names = "xdpe132g5h", "xdpe1a2g5b";
 		node_label = "XDP_E132G5H";
 		bus = <&i2c1>;
+		zephyr,deferred-init;
 	};
 
 	mp2891: mp2891@30 {
@@ -190,12 +194,14 @@
 		vout_scale = <1 1>;
 		iout_scale = <1 1>;
 		node_label = "MP2891";
+		zephyr,deferred-init;
 	};
 
 	adc121c027: adc121c027@51 {
 		compatible = "ti,adc121c027";
 		reg = <0x51>;
 		node_label = "ADC121C027";
+		zephyr,deferred-init;
 	};
 
 	qsfp0: qsfp0@50 {
@@ -207,6 +213,7 @@
 		modprsnt-gpios = <&gpio1 1 GPIO_ACTIVE_LOW>;
 		int-gpios = <&gpio1 0 GPIO_ACTIVE_LOW>;
 		cpld-reset-regs = <0x4 0x4>;
+		zephyr,deferred-init;
 	};
 
 	qsfp1: qsfp1@50 {
@@ -218,6 +225,7 @@
 		modprsnt-gpios = <&gpio2 5 GPIO_ACTIVE_LOW>;
 		int-gpios = <&gpio3 1 GPIO_ACTIVE_LOW>;
 		cpld-reset-regs = <0x4 0x8>;
+		zephyr,deferred-init;
 	};
 };
 
diff --git a/boards/hl/gaudi2/hl_gaudi2_mgmt_palad.dts b/boards/hl/gaudi2/hl_gaudi2_mgmt_palad.dts
index eeefbf7..2321a6e 100644
--- a/boards/hl/gaudi2/hl_gaudi2_mgmt_palad.dts
+++ b/boards/hl/gaudi2/hl_gaudi2_mgmt_palad.dts
@@ -112,6 +112,7 @@
 			remote_therm = <85 80>;
 			hysteresis= <60>;
 			node_label = "TMP464";
+			zephyr,deferred-init;
 		};
 
 		machxox: machxox@25 {
@@ -133,18 +134,21 @@
 			compatible = "ir,xdpp1100";
 			reg = <0x41>;
 			node_label = "XDPP_1100";
+			zephyr,deferred-init;
 		};
 
 		xdpe132g5h: xdpe132g5h@30 {
 			compatible = "ir,xdpe132g5h";
 			reg = <0x30>;
 			node_label = "XDP_E132G5H";
+			zephyr,deferred-init;
 		};
 
 		adc121c027: adc121c027@51 {
 			compatible = "ti,adc121c027";
 			reg = <0x51>;
 			node_label = "ADC121C027";
+			zephyr,deferred-init;
 		};
 
 		gpio_ex_2: ds4520@56 {
diff --git a/boards/hl/gaudi2/hl_gaudi2_preboot_asic.dts b/boards/hl/gaudi2/hl_gaudi2_preboot_asic.dts
index 7206f12..c452c6f 100644
--- a/boards/hl/gaudi2/hl_gaudi2_preboot_asic.dts
+++ b/boards/hl/gaudi2/hl_gaudi2_preboot_asic.dts
@@ -61,6 +61,7 @@
 		compatible = "ir,xdpp1100";
 		reg = <0x41>;
 		node_label = "XDPP_1100";
+		zephyr,deferred-init;
 	};
 
 	xdpe132g5h: xdpe132g5h@30 {
@@ -68,6 +69,7 @@
 		reg = <0x30>, <0x58>;
 		reg-names = "xdpe132g5h", "xdpe1a2g5b";
 		node_label = "XDP_E132G5H";
+		zephyr,deferred-init;
 	};
 
 	mp2891: mp2891@30 {
@@ -76,6 +78,7 @@
 		vout_scale = <1 1>;
 		iout_scale = <1 1>;
 		node_label = "MP2891";
+		zephyr,deferred-init;
 	};
 };
 
diff --git a/drivers/led/led_hl_cpld.c b/drivers/led/led_hl_cpld.c
index d4ea17a..d619216 100644
--- a/drivers/led/led_hl_cpld.c
+++ b/drivers/led/led_hl_cpld.c
@@ -101,11 +101,6 @@
 {
 	const struct led_cpld_config *config = DEV_CFG(dev);
 
-	if (!hl_is_sfp_present()) {
-		LOG_DBG("Not a PCIe card");
-		return -EINVAL;
-	}
-
 	if (config->i2c_dev == NULL) {
 		LOG_ERR("Failed to get I2C device");
 		return -EINVAL;
diff --git a/drivers/misc/hl_sfp/hl_sfp.c b/drivers/misc/hl_sfp/hl_sfp.c
index 802bae5..828875d 100644
--- a/drivers/misc/hl_sfp/hl_sfp.c
+++ b/drivers/misc/hl_sfp/hl_sfp.c
@@ -251,9 +251,6 @@
 	struct hl_sfp_dev_data *data = dev->data;
 	int err;
 
-	if (!hl_is_sfp_present())
-		return 0;
-
 	err = gpio_pin_configure_dt(&config->lpmod_gpio, GPIO_OUTPUT_INACTIVE);
 	if (err) {
 		LOG_ERR("%s failed to configure %s pin", config->label, "lpmod_gpio");
diff --git a/drivers/sensor/adc121c027/adc121c027.c b/drivers/sensor/adc121c027/adc121c027.c
index 970d7b1..64712f9 100644
--- a/drivers/sensor/adc121c027/adc121c027.c
+++ b/drivers/sensor/adc121c027/adc121c027.c
@@ -280,9 +280,6 @@
 	const struct adc121c027_config *cfg = DEV_CFG(dev);
 	union adc121c027_reg_conv_result raw_sample;
 
-	if (!hl_is_adc121c027_present())
-		return -EIO;
-
 	drv_data->i2c = cfg->i2c_bus_dev;
 	if (drv_data->i2c == NULL) {
 		LOG_ERR("adc121 i2c_bind fail");
diff --git a/drivers/sensor/mp2891/mp2891.c b/drivers/sensor/mp2891/mp2891.c
index 19946df..4f3cd6a 100644
--- a/drivers/sensor/mp2891/mp2891.c
+++ b/drivers/sensor/mp2891/mp2891.c
@@ -246,11 +246,6 @@
 	uint8_t buf[SMB_MAXBLOCKSIZE];
 	uint8_t count = 0;
 
-	if (!hl_is_mp2891_present()) {
-		LOG_DBG("Unsupported device on this board");
-		return -1;
-	}
-
 	LOG_INF("MP2891 device enumeration started");
 
 	data->client.dev = cfg->client_dev;
diff --git a/drivers/sensor/ti/ina23x/ina236.c b/drivers/sensor/ti/ina23x/ina236.c
index 4430fa2..d97e82e 100644
--- a/drivers/sensor/ti/ina23x/ina236.c
+++ b/drivers/sensor/ti/ina23x/ina236.c
@@ -17,8 +17,6 @@
 
 LOG_MODULE_REGISTER(INA236, CONFIG_SENSOR_LOG_LEVEL);
 
-#define INA236_12V_ADDR			0x42
-
 /**
  * @brief Macro used to convert milli vot to volt
  */
@@ -337,11 +335,6 @@
 
 	if (hl_is_sim_mode()) {
 		LOG_INF("%s skipped for simulation", __func__);
-		return 1;
-	}
-
-	if (config->bus.addr == INA236_12V_ADDR && !hl_is_ina236_0_present()) {
-		LOG_DBG("INA236 %x is not present in this board", config->bus.addr);
 		return 1;
 	}
 
diff --git a/drivers/sensor/tmp432/tmp432.c b/drivers/sensor/tmp432/tmp432.c
index a609791..0f94cff 100644
--- a/drivers/sensor/tmp432/tmp432.c
+++ b/drivers/sensor/tmp432/tmp432.c
@@ -428,11 +428,6 @@
 	struct tmp432_data *drv_data = DEV_DATA(dev);
 	uint8_t val;
 
-	if (!hl_is_tmp432_present()) {
-		LOG_DBG("Unsupported device on this board");
-		return -1;
-	}
-
 	LOG_INF("TMP432 %d Enumeration started", cfg->id);
 
 	drv_data->i2c = cfg->dev;
diff --git a/drivers/sensor/tmp464/tmp464.c b/drivers/sensor/tmp464/tmp464.c
index e7d5cbd..286ab80 100644
--- a/drivers/sensor/tmp464/tmp464.c
+++ b/drivers/sensor/tmp464/tmp464.c
@@ -774,11 +774,6 @@
 	const struct tmp464_config *cfg = DEV_CFG(dev);
 	uint16_t val;
 
-	if (!hl_is_tmp464_present()) {
-		LOG_DBG("Unsupported device on this board");
-		return -1;
-	}
-
 	drv_data->i2c = cfg->i2c_dev;
 	if (drv_data->i2c == NULL) {
 		TMP464_ERR("No %s dev", cfg->i2c_bus);
diff --git a/drivers/sensor/xdpe132g5h/xdpe132g5h.c b/drivers/sensor/xdpe132g5h/xdpe132g5h.c
index 9c0018b..b7a9479 100644
--- a/drivers/sensor/xdpe132g5h/xdpe132g5h.c
+++ b/drivers/sensor/xdpe132g5h/xdpe132g5h.c
@@ -761,11 +761,6 @@
 	uint8_t buf[SMB_MAXBLOCKSIZE];
 	uint8_t count = 0;
 
-	if (!hl_is_xdpe_present()) {
-		LOG_DBG("Unsupported device on this board");
-		return -1;
-	}
-
 	memset(buf, 0x0, SMB_MAXBLOCKSIZE);
 	LOG_INF("XDPE132G5H device enumeration started");
 
diff --git a/drivers/sensor/xdpp1100/xdpp1100.c b/drivers/sensor/xdpp1100/xdpp1100.c
index a077dd3..dcdd7b7 100644
--- a/drivers/sensor/xdpp1100/xdpp1100.c
+++ b/drivers/sensor/xdpp1100/xdpp1100.c
@@ -358,11 +358,6 @@
 	bool init_prot = DT_INST_PROP(0, init_prot);
 	int ret;
 
-	if (!hl_is_xdpp1100_present()) {
-		LOG_DBG("Unsupported device on this board");
-		return -1;
-	}
-
 	LOG_INF("XDPP1100 device enumeration started");
 
 	if (hl_is_sim_mode()) {
diff --git a/dts/arc/hl_gaudi3.dtsi b/dts/arc/hl_gaudi3.dtsi
index d10af7b..a4d22a4 100644
--- a/dts/arc/hl_gaudi3.dtsi
+++ b/dts/arc/hl_gaudi3.dtsi
@@ -555,6 +555,7 @@
 				node_label = "XDP_E132G5H_0";
 				status = "disabled";
 				bus = <&i2c1>;
+				zephyr,deferred-init;
 			};
 
 			xdpe132g5h_1: xdpe132g5h@33 {
@@ -563,6 +564,7 @@
 				node_label = "XDP_E132G5H_1";
 				status = "disabled";
 				bus = <&i2c1>;
+				zephyr,deferred-init;
 			};
 
 			xdpe132g5h_2: xdpe132g5h@34 {
@@ -571,6 +573,7 @@
 				node_label = "XDP_E132G5H_2";
 				status = "disabled";
 				bus = <&i2c1>;
+				zephyr,deferred-init;
 			};
 
 			xdpp1100: xdpp1100@31 {
@@ -579,6 +582,7 @@
 				node_label = "XDPP_1100";
 				status = "disabled";
 				bus = <&i2c1>;
+				zephyr,deferred-init;
 			};
 
 			qsfp0: qsfp0@50 {
@@ -591,6 +595,7 @@
 				int-gpios = <&d1_gpio1 4 GPIO_ACTIVE_LOW>;
 				cpld-reset-regs = <0x8 0x8>;
 				status = "disabled";
+				zephyr,deferred-init;
 			};
 
 			qsfp1: qsfp1@50 {
@@ -603,6 +608,7 @@
 				int-gpios = <&d1_gpio1 8 GPIO_ACTIVE_LOW>;
 				cpld-reset-regs = <0x8 0x10>;
 				status = "disabled";
+				zephyr,deferred-init;
 			};
 		};
 
diff --git a/dts/arc/hl_gaudi3_mgmt.dtsi b/dts/arc/hl_gaudi3_mgmt.dtsi
index 518cd8d..4166ab5 100644
--- a/dts/arc/hl_gaudi3_mgmt.dtsi
+++ b/dts/arc/hl_gaudi3_mgmt.dtsi
@@ -309,6 +309,7 @@
 		led-ctrl-reg = <0x0a>;
 		node_label = "HL338_LED_CPLD";
 		status = "okay";
+		zephyr,deferred-init;
 
 		led0: led_0 {
 			node_label = "QSFP1_LED_Y";
@@ -378,6 +379,7 @@
 		reg = <0x48>;
 		node_label = "TMP464_0";
 		status = "disabled";
+		zephyr,deferred-init;
 	};
 
 	tmp464_1: tmp464@49 {
@@ -385,6 +387,7 @@
 		reg = <0x49>;
 		node_label = "TMP464_1";
 		status = "disabled";
+		zephyr,deferred-init;
 	};
 
 	tmp432_0: tmp432@4c {
@@ -392,6 +395,7 @@
 		reg = <0x4C>;
 		node_label = "TMP432_0";
 		status = "disabled";
+		zephyr,deferred-init;
 	};
 
 	tmp432_1: tmp432@4d {
@@ -399,6 +403,7 @@
 		reg = <0x4D>;
 		node_label = "TMP432_1";
 		status = "disabled";
+		zephyr,deferred-init;
 	};
 
 	/* 12V Power meter */
@@ -409,6 +414,7 @@
 		max_current = <6000>;
 		rshunt = <5>;
 		status = "disabled";
+		zephyr,deferred-init;
 	};
 
 	/* 3.3V Power meter */
